mlflow/utils/requirements_utils.py

```diff
@@ -76,11 +90,13 @@ def _parse_requirements(requirements_file):
     for line in lines:
         if _is_requirements_file(line):
             req_file = line.split(maxsplit=1)[1]
-            abs_path = (
-                req_file
-                if os.path.isabs(req_file)
-                else os.path.join(os.path.dirname(requirements_file), req_file)
-            )
-            yield from _parse_requirements(abs_path)
+            # If `req_file` is an absolute path, `os.path.join` returns `req_file`:
+            # https://docs.python.org/3/library/os.path.html#os.path.join
+            abs_path = os.path.join(os.path.dirname(requirements_file), req_file)
+            yield from _parse_requirements(abs_path, is_constraint=False)
+        elif _is_constraints_file(line):
+            req_file = line.split(maxsplit=1)[1]
+            abs_path = os.path.join(os.path.dirname(requirements_file), req_file)
+            yield from _parse_requirements(abs_path, is_constraint=True)
         else:
-            yield line
+            yield _Requirement(line, is_constraint)
```

Comment:
This is the main change of this PR:

1. I added the elif branch to handle contraints files
2. Instead of returning a string, return a `_Requirement` instance which allows us to classify constraints.

Reply:
A large portion of the diff comes from refactoring and utilities that I newly introduced.